### PR TITLE
Grafana app platform: an aggregator cmd and package

### PR DESCRIPTION
**Pull request by charandas**
**What is this feature?**

Introduces a pkg/cmd for our version of kube-aggregator (similar to kube-apiserver). This one focuses on minimum requirements, gets rid of dependency on core V1 APIs and uses `externalname.grafana.app` for simple DNS reference to locally running aggregated services (extension-apiservers).

**Why do we need this feature?**

To be able to run Grafana effectively in the cloud, we need to break the monolith into individual API servers by group/version that each of them handles.

**Who is this feature for?**

Grafana App Platform

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/aggregator/aggregator.go [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+		if err := b.InstallSchema(aggregatorscheme.Scheme); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *AggregatorServerOptions) Config(codecs serializer.CodecFactory) (*genericapiserver.RecommendedConfig, error) {
+	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts(
+		"localhost", o.AlternateDNS, []net.IP{netutils.ParseIPSloppy("127.0.0.1")},
```
Comment (DanCech) [inline pkg/aggregator/aggregator.go:98]:
```suggestion
		"localhost", o.AlternateDNS, []net.IP{net.IPv4(127, 0, 0, 1)},
```

File: pkg/aggregator/aggregator.go [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+		apiregistrationClient.ApiregistrationV1(),
+		nil,
+		(func() ([]byte, []byte))(nil),
+		completedConfig.ExtraConfig.ServiceResolver,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	aggregatorServer.GenericAPIServer.AddPostStartHookOrDie("apiservice-status-override-available-controller", func(context genericapiserver.PostStartHookContext) error {
```
Comment (DanCech) [inline pkg/aggregator/aggregator.go:344]: since this function returns an error, we should use `AddPostStartHook` and return any error to our caller.

File: pkg/aggregator/aggregator.go [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+		if g == nil || len(g.PrioritizedVersions) < 1 {
+			continue
+		}
+		err = aggregatorServer.GenericAPIServer.InstallAPIGroup(g)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	time.Sleep(time.Second * 10)
```
Comment (DanCech) [inline pkg/aggregator/aggregator.go:371]: ?

File: pkg/apis/service/v0alpha1/register.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package v0alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name used in this package
+const GroupName = "service.grafana.app"
```
Comment (DanCech) [inline pkg/apis/service/v0alpha1/register.go:10]: how is this different from `GROUP`?

File: pkg/cmd/grafana/main.go [hunk 5]
```diff
@@ -45,6 +45,19 @@ func main() {
 					return nil
 				},
 			},
+			gsrv.ServerCommand(version, commit, enterpriseCommit, buildBranch, buildstamp),
+			{
+				// The kubernetes standalone apiserver service runner
+				Name:  "aggregator",
+				Usage: "run a standalone api service (experimental)",
```
Comment (DanCech) [inline pkg/cmd/grafana/main.go:52]: we should update this

File: playlist.yaml
Comment (ryantxu) [inline playlist.yaml:1]: this can be removed (a few exists in pkg/tests/apis/playlist/testdata)

File: pkg/apis/service/v0alpha1/register.go
Comment (ryantxu) [inline pkg/apis/service/v0alpha1/register.go:1]: is this file needed?  it seems like a duplicate of  https://github.com/grafana/grafana/pull/79948/files#diff-edb9a53dba9403351a6aa5cc63611b77608efe62ac019ba5e15a2ea65b163e37

File: pkg/aggregator/aggregator.go (hunk 3 above)
Comment (charandas) [reply pkg/aggregator/aggregator.go:373]: Oops, removing it now

File: pkg/apis/service/v0alpha1/register.go
Comment (charandas) [reply pkg/apis/service/v0alpha1/register.go:1]: Not sure. Its standard for codegen to dump one for individual versions in addition to the one up top.

File: pkg/apis/service/v0alpha1/register.go (hunk 4 above)
Comment (charandas) [reply pkg/apis/service/v0alpha1/register.go:10]: Looks like the same value. I would let @toddtreece comment. He worked on the `externalname` part and would be most aware of what is codegen here and what's authored by us.

Comment (toddtreece) [reply pkg/apis/service/v0alpha1/register.go:10]: i think this is just a byproduct of merging the k8s pattern for registration with the grafana pattern. this one can be removed and replaced with `GROUP` where appropriate

File: pkg/apis/service/v0alpha1/register.go
Comment (toddtreece) [reply pkg/apis/service/v0alpha1/register.go:1]: @charandas and i were just looking at this, and it looks like k8s codegen relies on some of these variables in `register.go`

File: pkg/aggregator/aggregator.go (hunk 2 above)
Comment (charandas) [reply pkg/aggregator/aggregator.go:346]: I think the `OrDie` variants do the same thing. They log the error using klog with fatal. We want to quit here anyway and this makes up for less code on top of the passed function handler. Would you agree?

File: pkg/services/sqlstore/migrator/dialect.go [hunk 6]
```diff
@@ -368,7 +368,7 @@ func (b *BaseDialect) InsertQuery(tableName string, row map[string]any) (string,
 	for col := range row {
 		keys = append(keys, col)
 	}
-	slices.Sort[string](keys)
+	slices.Sort(keys)
```
Comment (ryantxu) [inline pkg/services/sqlstore/migrator/dialect.go:371]: unrelated, but seems OK 🤷🏻

File: pkg/apis/service/v0alpha1/zz_generated.openapi.go [hunk 7]
```diff
@@ -0,0 +1,2605 @@
...
+		"k8s.io/apimachinery/pkg/apis/meta/v1.Condition":                        schema_pkg_apis_meta_v1_Condition(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.CreateOptions":                    schema_pkg_apis_meta_v1_CreateOptions(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.DeleteOptions":                    schema_pkg_apis_meta_v1_DeleteOptions(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.Duration":                         schema_pkg_apis_meta_v1_Duration(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.FieldsV1":                         schema_pkg_apis_meta_v1_FieldsV1(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.GetOptions":                       schema_pkg_apis_meta_v1_GetOptions(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.GroupKind":                        schema_pkg_apis_meta_v1_GroupKind(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.GroupResource":                    schema_pkg_apis_meta_v1_GroupResource(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.GroupVersion":                     schema_pkg_apis_meta_v1_GroupVersion(ref),
+		"k8s.io/apimachinery/pkg/apis/meta/v1.GroupVersionForDiscovery":         schema_pkg_apis_meta_v1_GroupVersionForDiscovery(ref),
```
Comment (ryantxu) [inline pkg/apis/service/v0alpha1/zz_generated.openapi.go:38]: where does all this come from?  If I change the update-codegen script to use be `service` I get a bunch of changes like:

image

and the openapi does not include lots of what is here

File: pkg/aggregator/resolver.go
Comment (ryantxu) [inline pkg/aggregator/resolver.go:1]: can we add a README.md in this folder explaining how this is different/modified from kube-aggregator?

IIUC, it is pretty much an exact copy *except* that we can resolve the service list differently?  

In kube-aggregator, external name is a property on service?  and for us it is its own resource? https://ibm.github.io/kubernetes-networking/services/externalname/

Comment (toddtreece) [reply pkg/aggregator/resolver.go:1]: > can we add a README.md in this folder explaining how this is different/modified from kube-aggregator?

that's a good call. i'll work on that today

File: pkg/apis/service/v0alpha1/zz_generated.openapi.go (hunk 7 above)
Comment (toddtreece) [reply pkg/apis/service/v0alpha1/zz_generated.openapi.go:38]: > where does all this come from?

i'm not sure i understand the scope of the question. do you mean the additional codegen output in general, or something specific to openapi?

Comment (ryantxu) [reply pkg/apis/service/v0alpha1/zz_generated.openapi.go:38]: I mean specifically, how to recreate it.  I tried the `hack/update-codegen.sh` script, but it produces different output

File: pkg/services/sqlstore/migrator/dialect.go (hunk 6 above)
Comment (charandas) [reply pkg/services/sqlstore/migrator/dialect.go:371]: Some go.mod upgrades that had this side-effect.

File: hack/make-aggregator-pki.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Comment (stephaniehingtgen) [inline hack/make-aggregator-pki.sh:1]: What would you think about adding a new folder to the `devenv` folder, with files relating to this project?

Comment (charandas) [reply hack/make-aggregator-pki.sh:1]: The [readme](https://github.com/grafana/grafana/blob/main/hack/README.md) in that folder explains the current reason for its existence. I piggy-backed on top of its existence and added my script there.

On an initial look at what devenv related targets do inside the Makefile, I see they are powered by docker-compose. As such, do you think, this is adjacent enough?

Also, might be better to wait to nail this down once we have more clarity on the aggregator usage in Grafana. It could potentially converge to not being a command.

File: pkg/aggregator/resolver.go
Comment (charandas) [reply pkg/aggregator/resolver.go:1]: I can work on this @toddtreece since you are out sick!

File: hack/make-aggregator-pki.sh (hunk 8 above)
Comment (stephaniehingtgen) [reply hack/make-aggregator-pki.sh:1]: Ah! I missed that readme, thanks for sharing that with me :)

Makes sense to keep it as-is, thanks for clarifying!

File: pkg/aggregator/README.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# aggregator
```
Comment (charandas) [inline pkg/aggregator/README.md:1]: Thoughts @ryantxu?

File: pkg/aggregator/aggregator.go [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+		apiServices = append(apiServices, apiService)
+	}
+
+	return apiServices
+}
+
+// createProxyTransport creates the dialer infrastructure to connect to the nodes.
+func createProxyTransport() *http.Transport {
+	var proxyDialerFn utilnet.DialFunc
+	// Proxying to pods and services is IP-based... don't expect to be able to verify the hostname
```
Comment (stephaniehingtgen) [inline pkg/aggregator/aggregator.go:492]: Nit: since it's going through the service IP, which will then route it to the pod - we may want to just state services, so that it isn't read as we are by-passing the services and going directly to the pods

```suggestion
	// Proxying to services is IP-based... don't expect to be able to verify the hostname
```

File: pkg/aggregator/aggregator.go [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+			continue
+		}
+		registration.AddAPIServiceToSyncOnStart(apiService)
+		apiServices = append(apiServices, apiService)
+	}
+
+	return apiServices
+}
+
+// createProxyTransport creates the dialer infrastructure to connect to the nodes.
```
Comment (stephaniehingtgen) [inline pkg/aggregator/aggregator.go:489]: Would this still be accurate? Nodes may be a bit overloaded in this context
```suggestion
// createProxyTransport creates the dialer infrastructure to connect to the api servers.
```

File: pkg/aggregator/README.md (hunk 9 above)
Comment (ryantxu) [reply pkg/aggregator/README.md:1]: looks great -- thanks.

**Review by ryantxu — APPROVED** [review]

**Review by ryantxu — APPROVED** [review]
